Block Extensions API module

Leo40Git opened the pull request:
This PR adds the Block Extensions module (and its containing blocks module).

- [x] Block Render Layer Map
- [x] Extended Block Settings
- [x] Extended Material Builder
- [x] Extendable Extra Block Data system

Leo40Git committed:
Init blocks module, add render layer map

Leo40Git committed:
Add license header (oops)

Leo40Git committed:
Add QuiltBlockSettings + QuiltMaterialBuilder

Extra data storage for QBS has been set up but is currently empty

Leo40Git commented:
Ready for review!

Leo40Git committed:
Add package-info to flag impl as internal

LambdAurora commented on library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/api/BlockRenderLayerMap.java:
```diff
@@ -0,0 +1,85 @@
+ * Provides methods to set the {@link RenderLayer} of blocks and fluids.
+ */
+@Environment(EnvType.CLIENT)
+public final class BlockRenderLayerMap {
```
Shouldn't this have an impl counterpart that is `ApiStatus.Internal` to avoid exposing the maps and the initialize method in the API?
And as this is client, this should be in a `client` subpackage.

Leo40Git replied:
Maps are `private`, so they're not actually exposed. The `initialize` method, however...

LambdAurora commented on library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/api/QuiltBlockSettings.java:
```diff
@@ -0,0 +1,254 @@
+		super(((AbstractBlockSettingsAccessor) settings).getMaterial(), ((AbstractBlockSettingsAccessor) settings).getMapColorProvider());
+
+		// [VanillaCopy] AbstractBlock.Settings#copy(AbstractBlock.Settings)
+		AbstractBlockSettingsAccessor thisAccessor = (AbstractBlockSettingsAccessor) this;
```
I would use `var` here to shorten the lines and the cast make it obvious which type the local is.
```suggestion
		var thisAccessor = (AbstractBlockSettingsAccessor) this;
```

LambdAurora commented on library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/api/QuiltBlockSettings.java:
```diff
@@ -0,0 +1,254 @@
+
+		// [VanillaCopy] AbstractBlock.Settings#copy(AbstractBlock.Settings)
+		AbstractBlockSettingsAccessor thisAccessor = (AbstractBlockSettingsAccessor) this;
+		AbstractBlockSettingsAccessor otherAccessor = (AbstractBlockSettingsAccessor) settings;
```
```suggestion
		var otherAccessor = (AbstractBlockSettingsAccessor) settings;
```

LambdAurora commented on library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/api/QuiltBlockSettings.java:
```diff
@@ -0,0 +1,254 @@
+
+	/* QUILT ADDTIIONS */
+
+	public QuiltBlockSettings luminance(int luminance) {
```
Javadoc would be welcomed.

LambdAurora commented on library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/api/QuiltMaterialBuilder.java:
```diff
@@ -0,0 +1,85 @@
+import net.minecraft.block.piston.PistonBehavior;
+import org.quiltmc.qsl.block.extensions.mixin.MaterialBuilderAccessor;
+
+public class QuiltMaterialBuilder extends Material.Builder {
```
Doc?

LambdAurora reviewed (commented):
Aside from the comments I made, it's nice overall, would love a package-info javadoc on the API package.

Leo40Git committed:
Varification

Co-authored-by: LambdAurora <aurora42lambda@gmail.com>

Leo40Git committed:
Remove impl details from BlockRenderLayerMap

Also move it to client package

Leo40Git committed:
An attempt at javadoc

Leo40Git committed:
a

Leo40Git committed:
Make onBuild use an event

Now we don't have to hardcode mining level stuff here like Fabric did!

Leo40Git committed:
package-info javadoc attempt

intellij won't let me import in package-info for JD, so you must suffer the wrath of qualifiers

Leo40Git commented:
Javadoc has been attempted.

LambdAurora reviewed (approved):
After re-reading I still can't get what's QuiltBlockInternals, but aside from that it's all good imo.

Leo40Git commented:
`QuiltBlockInternals` is for future work concerning extra block data. It's empty for now, will have more added onto it later.

Also I should probably make `ExtraData` extensible somehow, hm...

Leo40Git committed:
Add missing license headers

Leo40Git committed:
Reimplement ExtraData to be more extensible

Leo40Git committed:
Expose getting data collection, too

i509VCB commented on library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/api/client/BlockRenderLayerMap.java:
```diff
@@ -0,0 +1,77 @@
+	 * @param fluids target fluids
+	 */
+	public static void put(RenderLayer layer, Fluid... fluids) {
+		for (Fluid fluid : fluids)
```
Please use braces here for for loops.

i509VCB commented on library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/api/data/BlockDataKey.java:
```diff
@@ -0,0 +1,52 @@
+/**
+ * Represents a key that can be used to get a value from an {@link ExtraBlockData} instance.
+ *
+ * @param <T> value type
```
doesn't the doclet set the method docs for a record component when using the `@param` tag?

Leo40Git replied:
AFAIK, no.

i509VCB commented on library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/api/data/BlockDataKey.java:
```diff
@@ -0,0 +1,52 @@
+	 * @param type value type
+	 * @param name name
+	 */
+	public BlockDataKey { }
```
I'd argue this is redundant?

Leo40Git replied:
This is the only way to add Javadoc to record constructors/getters, AFAIK.

i509VCB commented on library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/api/data/ExtraBlockData.java:
```diff
@@ -0,0 +1,127 @@
+		/**
+		 * Creates a new builder.
+		 */
+		public Builder() {
```
Wouldn't it make more sense to have a `builder` method on `ExtraBlockData` and make this constructor private?

i509VCB commented on library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/api/data/ExtraBlockData.java:
```diff
@@ -0,0 +1,127 @@
+	/**
+	 * Invoked to compute an {@code ExtraBlockData} collection for a specific block.
+	 */
+	public interface OnBuild {
```
I need to think about this a bitr

Leo40Git committed:
Changes requested by i5

Add some crosslinks to data javadoc

BasiqueEvangelist commented on library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/api/QuiltMaterialBuilder.java:
```diff
@@ -0,0 +1,99 @@
+		return this;
+	}
+
+	@Override
+	public Material build() {
+		return super.build();
+	}
```
Is this needed, considering you can just use the superclass's `build` method?

Leo40Git replied:
Oops, accidentally overrode this. You'd think an IDE would say something about overriding a method just to call `super`, huh?

BasiqueEvangelist commented on library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/api/client/BlockRenderLayerMap.java:
```diff
@@ -0,0 +1,77 @@
+public final class BlockRenderLayerMap {
+	private BlockRenderLayerMap() { }
+
+	/**
+	 * Sets the render layer of the specified block.
+	 *
+	 * @param block target block
+	 * @param layer new render layer
+	 */
+	public static void put(Block block, RenderLayer layer) {
+		BlockRenderLayerMapImpl.put(block, layer);
+	}
```
Is the single block overload of this method needed, considering you can just use the multi block overload anyway?

BasiqueEvangelist commented on library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/api/client/BlockRenderLayerMap.java:
```diff
@@ -0,0 +1,77 @@
+		}
+	}
+
+	/**
+	 * Sets the render layer of the specified fluid.
+	 *
+	 * @param fluid target fluid
+	 * @param layer new render layer
+	 */
+	public static void put(Fluid fluid, RenderLayer layer) {
+		BlockRenderLayerMapImpl.put(fluid, layer);
+	}
```
Is the single fluid overload of this method needed, considering you can just use the multi fluid overload anyway?

BasiqueEvangelist commented on library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/api/data/ExtraBlockData.java:
```diff
@@ -0,0 +1,133 @@
+	 * Invoked to compute an {@code ExtraBlockData} collection for a specific block.
+	 */
+	public interface OnBuild {
+		ArrayEvent<OnBuild> EVENT = ArrayEvent.create(OnBuild.class, callbacks -> (block, settings, builder) -> {
+			for (OnBuild callback : callbacks)
+				callback.append(block, settings, builder);
+		});
```
Shouldn't this event be parameter invoking?

Leo40Git committed:
Apply code review

BlockRenderLayerMap: Remove single block/fluid overloads
ExtraBlockData.OnBuild: Make it invoke parameter (block)
QuiltMaterialBuilder: Remove useless build() override

QuiltBlockSettings: Add of(Material, Function<BlockState, MapColor>)

Leo40Git committed:
De-recordify BlockDataKey

Apparently trying to override a getter for a generic record component just... doesn't work

BasiqueEvangelist commented on library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/api/client/package-info.java:
```diff
@@ -0,0 +1,15 @@
+ * <h3>{@linkplain org.quiltmc.qsl.block.extensions.api.client.BlockRenderLayerMap Block render layer map}</h3>
+ * <ul>
+ *     <li>Provides methods for setting the render layer of
+ *     {@linkplain org.quiltmc.qsl.block.extensions.api.client.BlockRenderLayerMap#put(net.minecraft.block.Block, net.minecraft.client.render.RenderLayer) blocks}
+ *     and {@linkplain org.quiltmc.qsl.block.extensions.api.client.BlockRenderLayerMap#put(net.minecraft.fluid.Fluid, net.minecraft.client.render.RenderLayer) fluids}.</li>
```
This should be removed since the single-block and single-fluid overloads were removed.

BasiqueEvangelist commented on library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/api/data/ExtraBlockData.java:
```diff
@@ -0,0 +1,137 @@
+		 * @param settings block settings
+		 * @param builder collection builder
+		 */
+		@ParameterInvokingEvent
```
The `ParameterInvokingEvent` annotation is used incorrectly here. You're supposed to put it on a field that contains an event or a method which *returns* an event.

> ```
>  * <h2>How to use this annotation</h2>
>  *
>  * An event which may invoke a parameter should first place the annotation on the field of the event instance or the
>  * method which returns an event instance.
> ```

i509VCB replied:
~~Actually leo was correct here. Maybe I need to clarify it goes on the field/method which returns the event instance~~

i509VCB replied:
Never mind I was wrong

Leo40Git committed:
0 - The Fool (it's me, I'm the fool)

client.package-info: Fix dead Javadoc links
ExtraBlockData.OnBuild: ParameterInvokingEvent annotation being placed on the wrong member

Leo40Git committed:
BlockDataKey: Change name to Identifier

Leo40Git committed:
Allow setting extra data in QuiltBlockSettings

BasiqueEvangelist commented on library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/impl/QuiltBlockInternals.java:
```diff
@@ -0,0 +1,52 @@
+
+		ExtraBlockData extraData = internals.qsl$getExtraData();
+		if (extraData == null) {
+			ExtraBlockData.Builder builder = ExtraBlockData.builder();
+			Map<BlockDataKey<?>, Object> map = internals.qsl$getSettingsMap();
+			if (map != null) {
+				copyFromMap(builder, map);
+			}
+			ExtraBlockData.OnBuild.EVENT.invoker().append(block, settings, builder);
+			internals.qsl$setExtraData(extraData = builder.build());
```
Couldn't you give the QuiltBlockSettings extra data map directly to the builder, instead of copying everything to the builder?

Leo40Git replied:
I don't think I understood this. This is pretty much the most direct way to get the data map into the builder while not exposing a way to create an invalid `ExtraBlockData` collection (I.E keys mapped to values with incompatible types).

BasiqueEvangelist replied:
How would an invalid `ExtraBlockData` be created here?

Leo40Git replied:
Exposing a constructor that allows passing the map into the data collection/builder directly could allow this to happen:
```java
BlockDataKey<Integer> a = new BlockDataKey<>(Integer.class, new Identifier("badmod", "hehe"));
HashMap<BlockDataKey<?>, Object> map = new HashMap<>();
map.put(a, "This is not an integer!");
ExtraBlockData.Builder builder = ExtraBlockData.builder(map); // this builder will now construct an invalid collection!
```

BasiqueEvangelist replied:
Couldn't the constructor be an implementation detail? Yes, it may cause issues if it's public API ~~you could just cast the BlockDataKey anyway~~, but I'm saying it should be used *here* specifically.

Leo40Git committed:
Split ExtraBlockData implementation from API

Extra block data is now computed eagerly rather than on-demand

Leo40Git committed:
Add missing license headers & javadoc

BasiqueEvangelist commented on library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/impl/QuiltBlockInternals.java:
```diff
@@ -0,0 +1,49 @@
+		return extraData;
+	}
+
+	@SuppressWarnings("unchecked")
+	private static void copyFromMap(ExtraBlockData.Builder builder, Map<BlockDataKey<?>, Object> map) {
+		for (Map.Entry<BlockDataKey<?>, Object> entry : map.entrySet()) {
+			builder.put((BlockDataKey<Object>) entry.getKey(), entry.getValue());
+		}
+	}
```
This isn't used anywhere now.

Leo40Git replied:
Removed.

BasiqueEvangelist commented on library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/impl/QuiltBlockInternals.java:
```diff
@@ -0,0 +1,49 @@
+
+		ExtraBlockData extraData = internals.qsl$getExtraData();
+		if (extraData == null) {
+			Map<BlockDataKey<?>, Object> map = internals.qsl$getSettingsMap();
```
Isn't this map only made if `QuiltBlockSettings.extraData` was ever called?

Leo40Git replied:
...yes, it indeed is. Oops.

BasiqueEvangelist commented on library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/api/data/ExtraBlockData.java:
```diff
@@ -0,0 +1,110 @@
+ * Represents a {@linkplain BlockDataKey key} to value collection used to store extra data for {@link Block}s.
+ */
+public interface ExtraBlockData {
+	/**
+	 * Constructs a new builder.
+	 *
+	 * @return new builder
+	 */
+	static Builder builder() {
+		return new ExtraBlockDataImpl.BuilderImpl(new HashMap<>());
+	}
```
Is this needed? The block's data won't be constructed by the user anyway, you can only add to the builder.

BasiqueEvangelist commented on library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/api/data/ExtraBlockData.java:
```diff
@@ -0,0 +1,110 @@
+		 */
+		<T> Builder put(BlockDataKey<T> key, T value);
+
+		/**
+		 * Builds the collection.
+		 *
+		 * @return new collection
+		 */
+		ExtraBlockData build();
```
Same as for `ExtraBlockData.builder`. Is this needed?

Leo40Git committed:
QuiltBlockInternals: Fix potential NPE

Also remove unused method

ExtraBlockData: Remove builder() and Builder.build() methods, since they're not needed

BasiqueEvangelist commented on library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/impl/ExtraBlockDataImpl.java:
```diff
@@ -0,0 +1,52 @@
+		}
+
+		public ExtraBlockData build() {
+			return new ExtraBlockDataImpl(new HashMap<>(values));
```
You can just pass the `values` map to the `ExtraBlockDataImpl` now, since the builder'll be consumed after build anyway.

Leo40Git committed:
The less map allocations, the better

Leo40Git committed:
QuiltBlockSettings: also copy extra block data

BasiqueEvangelist commented on library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/api/QuiltBlockSettings.java:
```diff
@@ -0,0 +1,308 @@
+		super(material, mapColor);
+	}
+
+	protected QuiltBlockSettings(AbstractBlock.Settings settings) {
+		super(((AbstractBlockSettingsAccessor) settings).getMaterial(), ((AbstractBlockSettingsAccessor) settings).getMapColorProvider());
+
+		// [VanillaCopy] AbstractBlock.Settings#copy(AbstractBlock.Settings)
+		var thisAccessor = (AbstractBlockSettingsAccessor) this;
+		var otherAccessor = (AbstractBlockSettingsAccessor) settings;
+
+		thisAccessor.setMaterial(otherAccessor.getMaterial());
+		this.hardness(otherAccessor.getHardness());
+		this.resistance(otherAccessor.getResistance());
+		this.collidable(otherAccessor.getCollidable());
+		thisAccessor.setRandomTicks(otherAccessor.getRandomTicks());
+		this.luminance(otherAccessor.getLuminance());
+		thisAccessor.setMapColorProvider(otherAccessor.getMapColorProvider());
+		this.sounds(otherAccessor.getSoundGroup());
+		this.slipperiness(otherAccessor.getSlipperiness());
+		this.velocityMultiplier(otherAccessor.getVelocityMultiplier());
+		thisAccessor.setDynamicBounds(otherAccessor.getDynamicBounds());
+		thisAccessor.setOpaque(otherAccessor.getOpaque());
+		thisAccessor.setIsAir(otherAccessor.getIsAir());
+		thisAccessor.setToolRequired(otherAccessor.isToolRequired());
+
+		// also copy extra data
+		var thisInternals = (BlockSettingsInternals) this;
+		var otherInternals = (BlockSettingsInternals) this;
+
+		Map<BlockDataKey<?>, Object> settingsMap = otherInternals.qsl$getSettingsMap();
+		if (settingsMap != null)
+			thisInternals.qsl$setSettingsMap(new HashMap<>(settingsMap));
+	}
```
Does this copy the other settings' loot table ID?

Leo40Git replied:
Nope. Neither Vanilla nor Fabric bother with it, I'll add it now.

BasiqueEvangelist replied:
This also doesn't copy the `jumpVelocityMultiplier` and all the various `*Predicate`s

Leo40Git committed:
also copy loot table Id

Leo40Git committed:
copy e v e r y t h i n g

Leo40Git committed:
ExtraBlockData.Builder: add putIfAbsent/computeIfAbsent methods

Leo40Git committed:
ExtraBlockData.OnBuild: make method name more unique

Leo40Git committed:
ExtraBlockData.Builder.computeIfAbsent: make JD clearer

Leo40Git committed:
BlockDataKey: remove name

BlockDataKeys are now to be compared via identity rather than equals/hashCode
Changed QuiltBlockSettings/QuiltBlockInternals to match

Leo40Git committed:
Remove conditional builder methods in favor of default value in key

todo: Rename BlockDataKey since it's not really a key anymore

LambdAurora commented on library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/api/data/BlockDataKey.java:
```diff
@@ -0,0 +1,73 @@
+ *
+ */
+@SuppressWarnings("ClassCanBeRecord") // this key is identity-based, so it can't be a record!
+public final class BlockDataKey<T> {
```
Bit worried about the identity-baseness of this class as you will need to get a hold of the key reference for your stuff.

Like if you had something where the type is Float, which is managed by an external library, you're forced to bring that library in your dev env to be able to use the key since it's identity-based.

Leo40Git replied:
This entire system is being replaced (as soon as I get back to my laptop and also sleep), so this PR will be on hiatus until the new system is done, which will be another QSL library.

Leo40Git commented:
This PR is reverting to draft mode until #29 is completed and merged.

Leo40Git commented:
actually I'm just gonna rewrite this, so uh
bye PR
